locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/NetworkingEffectors.java

```diff
@@ -0,0 +1,90 @@
+import org.apache.brooklyn.util.core.BrooklynNetworkUtils;
+import org.apache.brooklyn.util.core.config.ConfigBag;
+import org.apache.brooklyn.util.net.Cidr;
+import org.apache.brooklyn.util.net.Networking;
+import org.jclouds.net.domain.IpPermission;
+import org.jclouds.net.domain.IpProtocol;
+
+import java.util.List;
+
+public class NetworkingEffectors {
+    // Intentionally not use CloudLocationConfig.INBOUND_PORTS to make richer syntax and rename it to differ it from the first in a ConfigBag
+    public static final ConfigKey<List<String>> INBOUND_PORTS_LIST = ConfigKeys.newConfigKey(new TypeToken<List<String>>() {}, "inbound.ports.list",
+            "Ports to open from the effector", ImmutableList.<String>of());
+    public static final ConfigKey<IpProtocol> INBOUND_PORTS_LIST_PROTOCOL = ConfigKeys.newConfigKey(new TypeToken<IpProtocol>() {}, "inbound.ports.list.protocol",
+            "Protocol for ports to open. Possible values: TCP, UDP, ICMP, ALL.", IpProtocol.TCP);
+
+    public static Effector<Iterable<IpPermission>> openPortsInSecurityGroupEffector() {
+        return Effectors.effector(EffectorResult.class, "openPortsInSecurityGroup")
+                .parameter(INBOUND_PORTS_LIST)
+                .parameter(INBOUND_PORTS_LIST_PROTOCOL)
+                .description("Open ports in Cloud Security Group. If called before machine location is provisioned, it will fail.")
+                .impl((EffectorBody)new OpenPortsInSecurityGroupBody())
+                .build();
+    }
+
+    private interface EffectorResult extends Iterable<IpPermission> {}
+
+    private static class OpenPortsInSecurityGroupBody extends EffectorBody<Iterable<IpPermission>> {
+        @Override
+        public Iterable<IpPermission> call(ConfigBag parameters) {
+            List<String> rawPortRules = parameters.get(INBOUND_PORTS_LIST);
+            IpProtocol ipProtocol = parameters.get(INBOUND_PORTS_LIST_PROTOCOL);
+            Preconditions.checkNotNull(ipProtocol);
+            Preconditions.checkNotNull(rawPortRules, "ports cannot be null");
+            MutableList.Builder<IpPermission> ipPermissionsBuilder = MutableList.builder();
+            for (Range<Integer> portRule : Networking.portRulesToRanges(rawPortRules).asRanges()) {
+                ipPermissionsBuilder.add(
+                        IpPermission.builder()
+                                .ipProtocol(ipProtocol)
+                                .fromPort(portRule.lowerEndpoint())
+                                .toPort(portRule.upperEndpoint())
+                                .cidrBlock(Cidr.UNIVERSAL.toString())
+                                .build());
+            }
+            JcloudsLocationSecurityGroupCustomizer customizer = JcloudsLocationSecurityGroupCustomizer.getInstance(entity());
+            Optional<Location> jcloudsMachineLocation = Iterables.tryFind(entity().getLocations(), Predicates.instanceOf(JcloudsMachineLocation.class));
```

Comment:
Use `Locations.getLocationsCheckingAncestors(locations, entity);` instead - the location could be declared on parent entities (SameServerEntity).